tower-http/src/validate_request.rs

```diff
@@ -0,0 +1,527 @@
+    pub fn custom(inner: S, validate: T) -> ValidateRequestHeader<S, T> {
+        Self { inner, validate }
+    }
+}
+
+impl<ReqBody, ResBody, S, V> Service<Request<ReqBody>> for ValidateRequestHeader<S, V>
+where
+    V: ValidateRequest<ReqBody, ResponseBody = ResBody>,
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+{
+    type Response = Response<ResBody>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, ResBody>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        match self.validate.validate(&mut req) {
+            Ok(_) => ResponseFuture::future(self.inner.call(req)),
+            Err(res) => ResponseFuture::invalid_header_value(res),
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`ValidateRequestHeader`].
+    pub struct ResponseFuture<F, B> {
+        #[pin]
+        kind: Kind<F, B>,
+    }
+}
+
+impl<F, B> ResponseFuture<F, B> {
+    fn future(future: F) -> Self {
+        Self {
+            kind: Kind::Future { future },
+        }
+    }
+
+    fn invalid_header_value(res: Response<B>) -> Self {
+        Self {
+            kind: Kind::Error {
+                response: Some(res),
+            },
+        }
+    }
+}
+
+pin_project! {
+    #[project = KindProj]
+    enum Kind<F, B> {
+        Future {
+            #[pin]
+            future: F,
+        },
+        Error {
+            response: Option<Response<B>>,
+        },
+    }
+}
+
+impl<F, B, E> Future for ResponseFuture<F, B>
+where
+    F: Future<Output = Result<Response<B>, E>>,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().kind.project() {
+            KindProj::Future { future } => future.poll(cx),
+            KindProj::Error { response } => {
+                /* Never panics unless polled after completion */
+                let response = response.take().unwrap();
+                Poll::Ready(Ok(response))
+            }
+        }
+    }
+}
+
+/// Trait for validating requests.
+pub trait ValidateRequest<B> {
+    /// The body type used for responses to unvalidated requests.
+    type ResponseBody;
+
+    /// Validate the request.
+    ///
+    /// If `Ok(())` is returned then the request is allowed through, otherwise not.
+    fn validate(&mut self, request: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>>;
+}
+
+impl<B, F, ResBody> ValidateRequest<B> for F
+where
+    F: FnMut(&mut Request<B>) -> Result<(), Response<ResBody>>,
+{
+    type ResponseBody = ResBody;
+
+    fn validate(&mut self, request: &mut Request<B>) -> Result<(), Response<Self::ResponseBody>> {
+        self(request)
+    }
+}
+
+/// Type that performs validation of the Accept header.
+pub struct AcceptHeader<ResBody> {
+    header_value: Mime,
```

Comment:
I think we should use `Arc<Mime>` since `Mime` might contain allocated collections like `String` and `Vec`. Just as a small perf optimization.

```suggestion
    header_value: Arc<Mime>,
```

Reply:
Ok, thanks. What kind of performance gains does this achieve?

Reply:
Services tend to be cloned a lot. So wrapping this in an `Arc` makes it cheaper to clone.